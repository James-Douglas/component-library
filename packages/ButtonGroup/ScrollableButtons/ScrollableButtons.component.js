import React, { useContext } from 'react';
import PropTypes from 'prop-types';
import { faAngleRight, faAngleLeft } from '@fortawesome/pro-regular-svg-icons';
import { ManorContext } from '@comparethemarketau/manor-provider';

import {
  ScrollableButtonsContainer, Tabs, ArrowIcon, ControlledButton,
} from './ScrollableButtons.style';
import useArrowControl from './hooks/useArrowControl';
import TabButton from './components/TabButton';

const ScrollableButtons = ({ trackingLabel, selectedId, buttons }) => {
  const { breakpoint } = useContext(ManorContext);

  const {
    showLeftArrow,
    showRightArrow,
    disableAllArrows,
    contentRef,
    leftArrowRef,
    rightArrowRef,
    handleLeftClick,
    handleRightClick,
    handleScroll,
    handleTabRefReady,
    handleTabMoveAround,
  } = useArrowControl(selectedId, trackingLabel);

  const defaultHref = '';

  // https://reactjs.org/docs/events.html
  // Starting with React 17, the onScroll event does not bubble in React.
  return (
    <ScrollableButtonsContainer>
      <ControlledButton
        direction="left"
        hidden={(breakpoint !== 'xs' || !showLeftArrow || disableAllArrows)}
        onClick={handleLeftClick}
        onKeyDown={handleLeftClick}
        role="button"
        tabIndex={0}
        ref={leftArrowRef}
      >
        <ArrowIcon icon={faAngleLeft} arrowdirection="Left" />
      </ControlledButton>
      <Tabs ref={contentRef} onScroll={handleScroll}>
        {
          buttons.map((button) => {
            const {
              label, id, handleClick, href,
            } = button;
            return (
              <TabButton
                key={`tab-${id}`}
                id={`tab-${id}`}
                selected={selectedId === id}
                label={label}
                href={href || defaultHref}
                handleClick={handleClick}
                onClick={() => { handleTabMoveAround(id); }}
                onRef={(tabRef) => handleTabRefReady(id, tabRef)}
              />
            );
          })
        }
      </Tabs>
      <ControlledButton
        direction="right"
        hidden={(breakpoint !== 'xs' || !showRightArrow || disableAllArrows)}
        onClick={handleRightClick}
        onKeyDown={handleRightClick}
        role="button"
        tabIndex={0}
        ref={rightArrowRef}
      >
        <ArrowIcon icon={faAngleRight} arrowdirection="Right" />
      </ControlledButton>
    </ScrollableButtonsContainer>
  );
};

ScrollableButtons.defaultProps = {
  selectedId: '',

};

ScrollableButtons.propTypes = {
  /** A descriptive label used in tracking user interactions with this component */
  trackingLabel: PropTypes.string.isRequired,
  /** The id of the focused button */
  selectedId: PropTypes.string,
  /** An array of buttons config */
  buttons: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.string,
    label: PropTypes.string,
    handleClick: PropTypes.func,
    href: PropTypes.string,
  })).isRequired,
};
export default ScrollableButtons;
