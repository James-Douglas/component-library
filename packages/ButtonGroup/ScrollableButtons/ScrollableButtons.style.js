import styled from 'styled-components';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

export const ScrollableButtonsContainer = styled.div`
  position: relative;
`;

export const Tabs = styled.ul`
  overflow: auto;
  scrollbar-width: none;
  -ms-overflow-style: none;
  background-color: ${({ theme }) => theme.colors.white};
  display: block;
  white-space: nowrap;
  padding-inline-start: 0;

  &::-webkit-scrollbar {
    display: none;
  };
`;

export const ArrowIcon = styled(FontAwesomeIcon)`
  color: ${({ theme }) => theme.colors.primary500};
  font-size: ${({ theme }) => theme.fontSize.base};
  margin-left: ${({ arrowdirection, theme }) => (arrowdirection === 'Left' ? theme.spacing[8] : '')};
  margin-right: ${({ arrowdirection, theme }) => (arrowdirection === 'Right' ? theme.spacing[8] : '')};
`;

export const ControlledButton = styled.div`
  cursor: pointer;
  display: ${({ hidden }) => (hidden ? 'none' : 'flex')};
  align-items: center;
  position: absolute;
  top: 0;
  bottom: 0;
  width: 4.8rem;
  ${({ direction }) => {
    switch (direction) {
      case 'left':
        return 'background: linear-gradient(91.03deg, #FFFFFF 35.83%, rgba(255, 255, 255, 0) 62.63%, rgba(255, 255, 255, 0) 72.01%); left: 0; justify-content: flex-start;';
      case 'right':
        return 'background: linear-gradient(271.03deg, #FFFFFF 21.83%, rgba(255, 255, 255, 0) 62.63%, rgba(255, 255, 255, 0) 72.01%); right: 0; justify-content: flex-end;';
      default:
        return '';
    }
  }};
`;
