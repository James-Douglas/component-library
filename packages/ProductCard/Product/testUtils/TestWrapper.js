import React from 'react';
import ProductProvider from '../../Context/ProductProvider';

const ProductTestWrapper = ({
  // eslint-disable-next-line react/prop-types
  children, product, onDetails = () => {}, onSelect = () => {},
}) => (
  <ProductProvider product={product} onDetails={onDetails} onSelect={onSelect}>
    { children }
  </ProductProvider>
);

export default ProductTestWrapper;
