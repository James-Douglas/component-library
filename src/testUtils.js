// test-utils.js
import {
  render, fireEvent, wait, act,
} from '@testing-library/react';
import ManorProvider from './providers/ManorProvider';

const customRender = (ui, options) => render(ui, { wrapper: ManorProvider, ...options });

// re-export everything
export * from '@testing-library/react';

// override render method
export {
  customRender as render,
  // export other used functions to appease eslint
  fireEvent,
  wait,
  act,
};
