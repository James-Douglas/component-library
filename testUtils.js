import React from 'react';
import {
  render, fireEvent, wait, act,
} from '@testing-library/react';
import ManorProvider from './packages/Provider/ManorProvider';

const ManorTestWrapper = ({ children }) => (
  <ManorProvider>
    {children}
  </ManorProvider>
);

const customRender = (ui, options) => render(ui, { wrapper: ManorTestWrapper, ...options });

// re-export everything
export * from '@testing-library/react';

// override render method
export {
  customRender as render,
  // export other used functions to appease eslint
  fireEvent,
  wait,
  act,
};
