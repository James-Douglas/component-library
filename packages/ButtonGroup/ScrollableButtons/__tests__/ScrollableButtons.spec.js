import React from 'react';
import { render } from '../../../../testUtils';
import ScrollableButtons from '../ScrollableButtons.component';

describe('ScrollableButtons', () => {
  it('renders with props', () => {
    const buttons = [
      {
        id: 'id-1',
        label: 'label1',
        handleClick: () => {},
        href: '#',
      },
      {
        id: 'id-2',
        label: 'label2',
        handleClick: () => {},
        href: '#',
      },
    ];

    const { container } = render(
      <ScrollableButtons trackingLabel="test-scrollable-buttons" selectedId="test" buttons={buttons} />,
    );

    expect(container).toMatchSnapshot();
  });
});
