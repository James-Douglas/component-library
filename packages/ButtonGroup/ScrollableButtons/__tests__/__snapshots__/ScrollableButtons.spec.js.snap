// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScrollableButtons renders with props 1`] = `
<div>
  <div
    class="sc-AxjAm hkuSct"
  >
    <div
      class="sc-AxhCb fTstWI"
      direction="left"
      hidden=""
      role="button"
      tabindex="0"
    >
      <svg
        aria-hidden="true"
        arrowdirection="Left"
        class="svg-inline--fa fa-angle-left sc-AxiKw hQrLWm"
        data-icon="angle-left"
        data-prefix="far"
        focusable="false"
        role="img"
        viewBox="0 0 192 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M4.2 247.5L151 99.5c4.7-4.7 12.3-4.7 17 0l19.8 19.8c4.7 4.7 4.7 12.3 0 17L69.3 256l118.5 119.7c4.7 4.7 4.7 12.3 0 17L168 412.5c-4.7 4.7-12.3 4.7-17 0L4.2 264.5c-4.7-4.7-4.7-12.3 0-17z"
          fill="currentColor"
        />
      </svg>
    </div>
    <ul
      class="sc-AxirZ bqraxx"
    >
      <li
        class="sc-AxgMl wmXNz"
      >
        <a
          class="MuiTypography-root MuiLink-root MuiLink-underlineNone MuiTypography-colorPrimary"
          href="#"
          style="font-family: \\"SourceSansPro\\", \\"Helvetica\\", \\"Arial\\"; color: rgb(11, 96, 183);"
        >
          <p
            class="sc-AxhUy lfOHbf sc-AxheI jfMvre"
          >
            label1
          </p>
        </a>
      </li>
      <li
        class="sc-AxgMl wmXNz"
      >
        <a
          class="MuiTypography-root MuiLink-root MuiLink-underlineNone MuiTypography-colorPrimary"
          href="#"
          style="font-family: \\"SourceSansPro\\", \\"Helvetica\\", \\"Arial\\"; color: rgb(11, 96, 183);"
        >
          <p
            class="sc-AxhUy lfOHbf sc-AxheI jfMvre"
          >
            label2
          </p>
        </a>
      </li>
    </ul>
    <div
      class="sc-AxhCb bmBWHl"
      direction="right"
      hidden=""
      role="button"
      tabindex="0"
    >
      <svg
        aria-hidden="true"
        arrowdirection="Right"
        class="svg-inline--fa fa-angle-right sc-AxiKw ldkiVq"
        data-icon="angle-right"
        data-prefix="far"
        focusable="false"
        role="img"
        viewBox="0 0 192 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M187.8 264.5L41 412.5c-4.7 4.7-12.3 4.7-17 0L4.2 392.7c-4.7-4.7-4.7-12.3 0-17L122.7 256 4.2 136.3c-4.7-4.7-4.7-12.3 0-17L24 99.5c4.7-4.7 12.3-4.7 17 0l146.8 148c4.7 4.7 4.7 12.3 0 17z"
          fill="currentColor"
        />
      </svg>
    </div>
  </div>
</div>
`;
